fpdf/table.py

```diff
@@ -268,23 +280,124 @@ def _render_table_cell(
-                align=text_align,
-                new_x="RIGHT",
-                new_y="TOP",
-                fill=fill,
-                markdown=self._markdown,
-                output=MethodReturnValue.PAGE_BREAK | MethodReturnValue.HEIGHT,
-                wrapmode=self._wrapmode,
-                **kwargs,
+
+        if cell.padding:
+            from fpdf.fpdf import get_padding_tuple
+
+            padding = get_padding_tuple(cell.padding)
+        else:
+            padding = self._padding
+
+        if cell.v_align:
+            v_align = cell.v_align
+        else:
+            v_align = self._v_align
+
+        # place cursor (required for images after images)
+        cell_widhts = [self._get_col_width(i, jj) for jj in range(j)]
+        cell_x = sum(cell_widhts)
+        self._fpdf.set_x(self._fpdf.l_margin + cell_x + self._gutter_width * j)
+
+        # render cell border and background
+
+        # if cell_height is defined, that means that we already know the size at which the cell will be rendered
+        # so we can draw the borders now
+        #
+        # If cell_height is None then we're still in the phase of calculating the height of the cell meaning that
+        # we do not need to render and borders yet.
+        if cell_height is not None:
+            with self._fpdf.use_font_face(style):
+                x1 = self._fpdf.x
+                y1 = self._fpdf.y
+                x2 = x1 + col_width
+                y2 = y1 + cell_height
+
+                self._fpdf._draw_box(
+                    x1, y1, x2, y2, border=self.get_cell_border(i, j), fill=fill
+                )
+
+        # render image
+
+        if cell.img:
+            x, y = self._fpdf.x, self._fpdf.y
+
+            # if cell_height is None or width is given then call image with h=0
+            # calling with h=0 means that the image will be rendered with an auto determined height
+            auto_height = cell.img_fill_width or cell_height is None
+
+            # apply padding
+            self._fpdf.x += padding[3]
+            self._fpdf.y += padding[0]
+
+            image = self._fpdf.image(
+                cell.img,
+                w=col_width - padding[1] - padding[3],
+                h=0 if auto_height else cell_height - padding[0] - padding[2],
+                keep_aspect_ratio=True,
             )
-        return page_break, max(img_height, cell_height)
+
+            img_height = image.rendered_height
+
+            if img_height + y > self._fpdf.page_break_trigger:
+                page_break_image = True
+
+            self._fpdf.set_xy(x, y)
+
+        # render text
+
+        if cell.text:
+            dy = 0
+            if cell_height is not None:
+                if v_align != AlignV.T:  # For Top we don't need to calculate the dy
+                    # first dry run to get the actual text height of the cell
+                    _, actual_text_height = self._fpdf.multi_cell(
```

Comment:
You approach here introduces an extra call to `FPDF.multi_cell()` for every cell in the table...
This is a bit costly and will slow down table rendering for everyone, even if they don't use `padding` or `v_align`!

Couldn't you compute this `dy` during the "first pass" made in `_get_row_layout_info()`?

In fact, during the "second pass"
(when `_render_table_cell()` is called from `_render_table_row()`),
shouldn't `actual_text_height` here always be equal to `row_height`?

Reply:
There is indeed probably a more efficient way to do this. Will look into that once everything else works.

Reply:
As you wish, but I think this is a structural problem, and the major feedback point I have regarding your PR currently.

Reply:
marked with #TODO

Reply:
> marked with #TODO

What does that mean?

Reply:
And done by adding the cell height info to RowLayoutInfo.